# Lab 7: Support Vector Machines (SVM)

---



## Importing the libraries

In [1]:
import pandas as pd
from sklearn.svm import SVC
from sklearn.model_selection import cross_val_score

## Importing the dataset

In [2]:
train_dataset = pd.read_excel("TrainingSet.xlsx")
test_dataset = pd.read_excel("TestingSet.xlsx")

In [3]:
train_dataset

leaf.length  leaf.width  flower.length  flower.width       plant
0            5.4         3.7            1.5           0.2     Arctica
1            4.8         3.4            1.6           0.2     Arctica
2            4.8         3.0            1.4           0.1     Arctica
3            4.3         3.0            1.1           0.1     Arctica
4            5.8         4.0            1.2           0.2     Arctica
..           ...         ...            ...           ...         ...
115          6.7         3.0            5.2           2.3  Carolinian
116          6.3         2.5            5.0           1.9  Carolinian
117          6.5         3.0            5.2           2.0  Carolinian
118          6.2         3.4            5.4           2.3  Carolinian
119          5.9         3.0            5.1           1.8  Carolinian

[120 rows x 5 columns]

In [4]:
test_dataset

leaf.length  leaf.width  flower.length  flower.width  plant
0           4.4         2.9            1.4           0.2    NaN
1           4.6         3.1            1.5           0.2    NaN
2           4.6         3.4            1.4           0.3    NaN
3           4.7         3.2            1.3           0.2    NaN
4           4.9         3.0            1.4           0.2    NaN
5           4.9         3.1            1.5           0.1    NaN
6           4.9         2.4            3.3           1.0    NaN
7           4.9         2.5            4.5           1.7    NaN
8           5.0         3.6            1.4           0.2    NaN
9           5.0         3.4            1.5           0.2    NaN
10          5.1         3.5            1.4           0.2    NaN
11          5.2         2.7            3.9           1.4    NaN
12          5.4         3.9            1.7           0.4    NaN
13          5.5         2.3            4.0           1.3    NaN
14          5.7         2.8            4.5           1.3    NaN
15          5.8         2.7            5.1           1.9    NaN
16          6.3         3.3            4.7           1.6    NaN
17          6.3         3.3            6.0           2.5    NaN
18          6.3         2.9            5.6           1.8    NaN
19          6.4         3.2            4.5           1.5    NaN
20          6.5         2.8            4.6           1.5    NaN
21          6.5         3.0            5.8           2.2    NaN
22          6.6         2.9            4.6           1.3    NaN
23          6.7         2.5            5.8           1.8    NaN
24          6.9         3.1            4.9           1.5    NaN
25          7.0         3.2            4.7           1.4    NaN
26          7.1         3.0            5.9           2.1    NaN
27          7.2         3.6            6.1           2.5    NaN
28          7.3         2.9            6.3           1.8    NaN
29          7.6         3.0            6.6           2.1    NaN

## Data preprocessing

In [5]:
X_train = train_dataset.iloc[:, :-1].values
y_train = train_dataset.iloc[:, -1].values

In [6]:
X_train.shape

(120, 4)

In [7]:
print(X_train)

[[5.4 3.7 1.5 0.2]
 [4.8 3.4 1.6 0.2]
 [4.8 3.  1.4 0.1]
 [4.3 3.  1.1 0.1]
 [5.8 4.  1.2 0.2]
 [5.7 4.4 1.5 0.4]
 [5.4 3.9 1.3 0.4]
 [5.1 3.5 1.4 0.3]
 [5.7 3.8 1.7 0.3]
 [5.1 3.8 1.5 0.3]
 [5.4 3.4 1.7 0.2]
 [5.1 3.7 1.5 0.4]
 [4.6 3.6 1.  0.2]
 [5.1 3.3 1.7 0.5]
 [4.8 3.4 1.9 0.2]
 [5.  3.  1.6 0.2]
 [5.  3.4 1.6 0.4]
 [5.2 3.5 1.5 0.2]
 [5.2 3.4 1.4 0.2]
 [4.7 3.2 1.6 0.2]
 [4.8 3.1 1.6 0.2]
 [5.4 3.4 1.5 0.4]
 [5.2 4.1 1.5 0.1]
 [5.5 4.2 1.4 0.2]
 [4.9 3.1 1.5 0.2]
 [5.  3.2 1.2 0.2]
 [5.5 3.5 1.3 0.2]
 [4.9 3.6 1.4 0.1]
 [4.4 3.  1.3 0.2]
 [5.1 3.4 1.5 0.2]
 [5.  3.5 1.3 0.3]
 [4.5 2.3 1.3 0.3]
 [4.4 3.2 1.3 0.2]
 [5.  3.5 1.6 0.6]
 [5.1 3.8 1.9 0.4]
 [4.8 3.  1.4 0.3]
 [5.1 3.8 1.6 0.2]
 [4.6 3.2 1.4 0.2]
 [5.3 3.7 1.5 0.2]
 [5.  3.3 1.4 0.2]
 [5.  2.  3.5 1. ]
 [5.9 3.  4.2 1.5]
 [6.  2.2 4.  1. ]
 [6.1 2.9 4.7 1.4]
 [5.6 2.9 3.6 1.3]
 [6.7 3.1 4.4 1.4]
 [5.6 3.  4.5 1.5]
 [5.8 2.7 4.1 1. ]
 [6.2 2.2 4.5 1.5]
 [5.6 2.5 3.9 1.1]
 [5.9 3.2 4.8 1.8]
 [6.1 2.8 4.  1.3]
 [6.3 2.5 4.

In [8]:
y_train.shape

(120,)

In [9]:
print(y_train)

['Arctica' 'Arctica' 'Arctica' 'Arctica' 'Arctica' 'Arctica' 'Arctica'
 'Arctica' 'Arctica' 'Arctica' 'Arctica' 'Arctica' 'Arctica' 'Arctica'
 'Arctica' 'Arctica' 'Arctica' 'Arctica' 'Arctica' 'Arctica' 'Arctica'
 'Arctica' 'Arctica' 'Arctica' 'Arctica' 'Arctica' 'Arctica' 'Arctica'
 'Arctica' 'Arctica' 'Arctica' 'Arctica' 'Arctica' 'Arctica' 'Arctica'
 'Arctica' 'Arctica' 'Arctica' 'Arctica' 'Arctica' 'Harlequin' 'Harlequin'
 'Harlequin' 'Harlequin' 'Harlequin' 'Harlequin' 'Harlequin' 'Harlequin'
 'Harlequin' 'Harlequin' 'Harlequin' 'Harlequin' 'Harlequin' 'Harlequin'
 'Harlequin' 'Harlequin' 'Harlequin' 'Harlequin' 'Harlequin' 'Harlequin'
 'Harlequin' 'Harlequin' 'Harlequin' 'Harlequin' 'Harlequin' 'Harlequin'
 'Harlequin' 'Harlequin' 'Harlequin' 'Harlequin' 'Harlequin' 'Harlequin'
 'Harlequin' 'Harlequin' 'Harlequin' 'Harlequin' 'Harlequin' 'Harlequin'
 'Harlequin' 'Harlequin' 'Carolinian' 'Carolinian' 'Carolinian'
 'Carolinian' 'Carolinian' 'Carolinian' 'Carolinian' 'Carolinian'
 '

In [10]:
X_test = test_dataset.iloc[:, :-1].values

In [11]:
X_test.shape

(30, 4)

In [12]:
print(X_test)

[[4.4 2.9 1.4 0.2]
 [4.6 3.1 1.5 0.2]
 [4.6 3.4 1.4 0.3]
 [4.7 3.2 1.3 0.2]
 [4.9 3.  1.4 0.2]
 [4.9 3.1 1.5 0.1]
 [4.9 2.4 3.3 1. ]
 [4.9 2.5 4.5 1.7]
 [5.  3.6 1.4 0.2]
 [5.  3.4 1.5 0.2]
 [5.1 3.5 1.4 0.2]
 [5.2 2.7 3.9 1.4]
 [5.4 3.9 1.7 0.4]
 [5.5 2.3 4.  1.3]
 [5.7 2.8 4.5 1.3]
 [5.8 2.7 5.1 1.9]
 [6.3 3.3 4.7 1.6]
 [6.3 3.3 6.  2.5]
 [6.3 2.9 5.6 1.8]
 [6.4 3.2 4.5 1.5]
 [6.5 2.8 4.6 1.5]
 [6.5 3.  5.8 2.2]
 [6.6 2.9 4.6 1.3]
 [6.7 2.5 5.8 1.8]
 [6.9 3.1 4.9 1.5]
 [7.  3.2 4.7 1.4]
 [7.1 3.  5.9 2.1]
 [7.2 3.6 6.1 2.5]
 [7.3 2.9 6.3 1.8]
 [7.6 3.  6.6 2.1]]


## Training the SVM (kernel = linear)

In [13]:
svm_linear = SVC(kernel = 'linear', random_state = 0)

In [14]:
svm_linear.fit(X_train, y_train)

SVC(kernel='linear', random_state=0)

In [15]:
accuracy_scores1 = cross_val_score(svm_linear, X_train, y_train, cv = 5)
print("Cross-validation Accuracy Scores:")
for i in range(5):
  print(f'cv = {i+1} : accuracy = {(accuracy_scores1[i])*100:.2f} %')
mean_accuracy1 = accuracy_scores1.mean()
print('\nMean Accuracy: ' + str(mean_accuracy1*100) + ' %')

Cross-validation Accuracy Scores:
cv = 1 : accuracy = 100.00 %
cv = 2 : accuracy = 95.83 %
cv = 3 : accuracy = 91.67 %
cv = 4 : accuracy = 100.00 %
cv = 5 : accuracy = 100.00 %

Mean Accuracy: 97.5 %


## Training the SVM (kernel = rbf)

In [16]:
svm_rbf = SVC(kernel = 'rbf', random_state = 0)

In [17]:
svm_rbf.fit(X_train, y_train)

SVC(random_state=0)

In [18]:
accuracy_scores2 = cross_val_score(svm_rbf, X_train, y_train, cv = 5)
print("Cross-validation Accuracy Scores:")
for i in range(5):
  print(f'cv = {i+1} : accuracy = {accuracy_scores2[i]*100:.2f} %')
mean_accuracy2 = accuracy_scores2.mean()
print(f'\nMean Accuracy: {mean_accuracy2*100:.2f} %')

Cross-validation Accuracy Scores:
cv = 1 : accuracy = 100.00 %
cv = 2 : accuracy = 91.67 %
cv = 3 : accuracy = 91.67 %
cv = 4 : accuracy = 100.00 %
cv = 5 : accuracy = 100.00 %

Mean Accuracy: 96.67 %


## Making the predictions

In [19]:
predictions1 = svm_linear.predict(X_test)

In [20]:
predictions1

array(['Arctica', 'Arctica', 'Arctica', 'Arctica', 'Arctica', 'Arctica',
       'Harlequin', 'Harlequin', 'Arctica', 'Arctica', 'Arctica',
       'Harlequin', 'Arctica', 'Harlequin', 'Harlequin', 'Carolinian',
       'Harlequin', 'Carolinian', 'Carolinian', 'Harlequin', 'Harlequin',
       'Carolinian', 'Harlequin', 'Carolinian', 'Harlequin', 'Harlequin',
       'Carolinian', 'Carolinian', 'Carolinian', 'Carolinian'],
      dtype=object)

In [21]:
test_dataset['plant'] = predictions1

In [22]:
test_dataset.to_excel('SVM kernel=linear Predictions.xlsx', index=False)

In [23]:
predictions2 = svm_rbf.predict(X_test)

In [24]:
predictions2

array(['Arctica', 'Arctica', 'Arctica', 'Arctica', 'Arctica', 'Arctica',
       'Harlequin', 'Harlequin', 'Arctica', 'Arctica', 'Arctica',
       'Harlequin', 'Arctica', 'Harlequin', 'Harlequin', 'Carolinian',
       'Harlequin', 'Carolinian', 'Carolinian', 'Harlequin', 'Harlequin',
       'Carolinian', 'Harlequin', 'Carolinian', 'Harlequin', 'Harlequin',
       'Carolinian', 'Carolinian', 'Carolinian', 'Carolinian'],
      dtype=object)

In [25]:
test_dataset['plant'] = predictions2

In [26]:
test_dataset.to_excel('SVM kernel=rbf Predictions.xlsx', index=False)

## Compariing the predictions

In [27]:
pred1 = pd.Series(predictions1, name = 'SVM (linear)')
pred2 = pd.Series(predictions2, name = 'SVM (rbf)')
comparison = pd.concat([pred1, pred2], axis = 1)
print(comparison)

   SVM (linear)   SVM (rbf)
0       Arctica     Arctica
1       Arctica     Arctica
2       Arctica     Arctica
3       Arctica     Arctica
4       Arctica     Arctica
5       Arctica     Arctica
6     Harlequin   Harlequin
7     Harlequin   Harlequin
8       Arctica     Arctica
9       Arctica     Arctica
10      Arctica     Arctica
11    Harlequin   Harlequin
12      Arctica     Arctica
13    Harlequin   Harlequin
14    Harlequin   Harlequin
15   Carolinian  Carolinian
16    Harlequin   Harlequin
17   Carolinian  Carolinian
18   Carolinian  Carolinian
19    Harlequin   Harlequin
20    Harlequin   Harlequin
21   Carolinian  Carolinian
22    Harlequin   Harlequin
23   Carolinian  Carolinian
24    Harlequin   Harlequin
25    Harlequin   Harlequin
26   Carolinian  Carolinian
27   Carolinian  Carolinian
28   Carolinian  Carolinian
29   Carolinian  Carolinian


## Conclusion

**SVM with Linear Kernel**

*   Cross-validation accuracy scores range from 91.67% to 100.00% across different folds.
*   Mean accuracy is calculated as 97.5% based on the cross-validation results.

**SVM with RBF Kernel**

*   Cross-validation accuracy scores range from 91.67% to 100.00% across different folds, similar to the linear SVM.
*   Mean accuracy is slightly lower compared to the linear SVM, calculated at 96.67% based on the cross-validation results.

**Comparison**

*   Both SVM classifiers perform very well on the plant classification task, with accuracies consistently above 90% across different folds.
*   The linear SVM shows a slightly higher mean accuracy (97.5%) compared to the RBF SVM (96.67%) based on cross-validation results.
*   It's notable that both SVM classifiers produce identical predictions for all instances in the test dataset, indicating robustness and generalizability in their predictions.



---

